### PR TITLE
Ide: Add command to close all tabs to the right

### PR DESCRIPTION
**Pull request by sandyarmstrong**
Identical behavior to VScode.

image

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs [hunk 1]
```diff
@@ -64,26 +67,39 @@ protected override void Run ()
...
-				.ToArray ();
+			var docs = new List<Document> ();
+			var dirtyDialogShown = false;
+			var startRemoving = except == null || !StartAfterException;
+
+			foreach (var doc in IdeApp.Workbench.Documents) {
+				if (((SdiWorkspaceWindow)doc.Window).TabControl == activeNotebook) {
+					if (except != null && doc.Window.ViewContent == except) {
+						startRemoving = true;
+						continue;
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: Won't this ignore the active document in CloseAll?

Comment (sandyarmstrong) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: I tested Close All, Close Except, Close Tab, Close To The Right, all with tab reordering. I think we're good. Thanks for checking!

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs [hunk 2]
```diff
@@ -92,15 +108,20 @@ protected override ViewContent GetDocumentException ()
 			return active == null ? null : active.Window.ViewContent;
 		}
 	}
-	
+
+	class CloseAllToTheRightHandler : CloseAllButThisHandler
+	{
+		protected override bool StartAfterException => true;
```
Comment (sevoku) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: What about enabling it only if there are any documents to the right?
We could add the same to `CloseAllButThisHandler` but without the `.SkipWhile()` part.

```suggestion
		protected override bool StartAfterException => true;
		
		
		protected override void Update (CommandInfo info)
		{
			var active = IdeApp.Workbench.ActiveDocument;
			if (active == null)
				return;

			var activeNotebook = ((SdiWorkspaceWindow)active.Window).TabControl;
			info.Enabled = IdeApp.Workbench.Documents.SkipWhile (doc => doc != active).Any (doc => doc != active && ((SdiWorkspaceWindow)doc.Window).TabControl == activeNotebook);
		}
```

Comment (sandyarmstrong) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: @sevoku I have mixed feelings about this, which is why I left it out. How often is `Update` called for menu items? I know it's a lot for always-visible toolbar commands.

VScode doesn't bother disabling for the right-most tab, for what it's worth.

Happy to add this if we're not worried about `Update` doing work.

Comment (netonjm) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: I think the update is only called when the popup is opened. You only need to check is if is the last element of the array.

Comment (sandyarmstrong) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: Last element of the array with matching `TabControl`, yeah. I'll update on Monday, thanks!

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs [hunk 3]
```diff
@@ -64,43 +67,106 @@ protected override void Run ()
...
+				info.Enabled = false;
+				return;
+			}
+
+			var activeNotebook = ((SdiWorkspaceWindow)activeDoc.Window).TabControl;
+
+			// Disable if right-most document in tab strip
+			for (int i = documents.Count - 1; i >= 0; i--) {
+				var doc = documents [i];
+				if (((SdiWorkspaceWindow)doc.Window).TabControl == activeNotebook) {
```
Comment (netonjm) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: we can remove this from the iteration, we get always the same value

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs [hunk 4]
```diff
@@ -64,43 +67,106 @@ protected override void Run ()
...
+
+			if (activeDoc == null) {
+				info.Enabled = false;
+				return;
+			}
+
+			var activeNotebook = ((SdiWorkspaceWindow)activeDoc.Window).TabControl;
+
+			// Disable if right-most document in tab strip
+			for (int i = documents.Count - 1; i >= 0; i--) {
```
[the remark this answers is not in the document]
Comment (sandyarmstrong) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: The last member of the `documents` list could be in a different tab strip.

[the remark this answers is not in the document]
Comment (netonjm) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: Sorry I was assuming we were checking only checking the documents from the tabstrip of the *selected tab*

[the remark this answers is not in the document]
Comment (sevoku) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: Hhm, I'm actually not sure, are the docs in `documents` sorted by their occurrence in the tab view?

[the remark this answers is not in the document]
Comment (sandyarmstrong) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/FileTabCommands.cs]: Yes.

**Comment by sandyarmstrong** [comment]
Can't merge because Mac tests are timing out still.

**Comment by sandyarmstrong** [comment]
This already merged.

**Pull request closed without merging**